server/mqtt_ex_test_test.go

```diff
@@ -0,0 +1,569 @@
+// Copyright 2024 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !skip_mqtt_tests
+// +build !skip_mqtt_tests
+
+package server
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"testing"
+
+	"github.com/nats-io/nuid"
+)
+
+type mqttExDial string
+
+type mqttExTarget struct {
+	servers  []*Server
+	clusters []*cluster
+	configs  []mqttExTestConfig
+	all      []mqttExDial
+}
+
+type mqttExTestConfig struct {
+	name string
+	pub  []mqttExDial
+	sub  []mqttExDial
+}
+
+func TestMQTTExCompliance(t *testing.T) {
+	if mqttexCLICommandPath == "" {
+		t.Skip(`"mqtt" command is not found in $PATH nor $MQTT_CLI. See https://hivemq.github.io/mqtt-cli/docs/installation/#debian-package for installation instructions`)
+	}
+
+	o := testMQTTDefaultOptions()
+	s := testMQTTRunServer(t, o)
+	o = s.getOpts()
+	defer testMQTTShutdownServer(s)
+
+	cmd := exec.Command(mqttexCLICommandPath, "test", "-V", "3", "-p", strconv.Itoa(o.MQTT.Port))
+
+	output, err := cmd.CombinedOutput()
+	t.Log(string(output))
+	if err != nil {
+		if exitError, ok := err.(*exec.ExitError); ok {
+			t.Fatalf("mqtt cli exited with error: %v", exitError)
+		}
+	}
+}
+
+func TestMQTTExRetainedMessages(t *testing.T) {
+	if mqttexTestCommandPath == "" {
+		t.Skip(`"mqtt-test" command is not found in $PATH.`)
+	}
+
+	for _, topo := range []struct {
+		name  string
+		makef func(testing.TB) *mqttExTarget
+	}{
+		{
+			name:  "single server",
+			makef: mqttExMakeServer,
+		},
+		{
+			name:  "server with leafnode",
+			makef: mqttExMakeServerWithLeafnode("HUBD", "LEAFD", true),
+		},
+		{
+			name:  "server with leafnode no domains",
+			makef: mqttExMakeServerWithLeafnode("", "", true),
+		},
+		{
+			name:  "server with leafnode no system account",
+			makef: mqttExMakeServerWithLeafnode("HUBD", "LEAFD", false),
+		},
+		{
+			name:  "cluster",
+			makef: mqttExMakeCluster(4, ""),
+		},
+		{
+			name:  "cluster with leafnode cluster",
+			makef: mqttExMakeClusterWithLeafnodeCluster("HUBD", "LEAFD", true),
+		},
+		{
+			name:  "cluster with leafnode cluster no system account",
+			makef: mqttExMakeClusterWithLeafnodeCluster("HUBD", "LEAFD", false),
+		},
+	} {
+		t.Run(topo.name, func(t *testing.T) {
+			target := topo.makef(t)
+			t.Cleanup(target.Shutdown)
+
+			// TODO (levb): if we do not pre-initialize the streams and
```

Comment:
NS: cc @derekcollison @kozlovic This test uncovered a rather inconsistent behavior in multi-domain scenarios. If there are 2 domains, A and B, and the MQTT JetStream assets are first initialized in A, B will only receive and replay retained messages after its own retained message assets are created. Touching the servers with an MQTT connection here ensures the success of the tests; otherwise A<->B fail 1/2 of the time.

Not quite sure if this is by design.

Reply:
I would think that if they have different domains, messages should not even be visible in the other domain, but I have to say I don't really understand that concept. What I notice is that you don't specify `js_domain: ` in the `mqtt{}` block of the config, but if I try to add those, then the test fails early since the lookup of a stream fails with a timeout since the JS APIs are rewritten with the domain name, like `$JS.HUBD.API.STREAM.INFO.$MQTT_sess` and for some reason, the server does not listen to those, even though you have specified `domain` in the JS config too. I don't really see in JS code where the server listens to that...